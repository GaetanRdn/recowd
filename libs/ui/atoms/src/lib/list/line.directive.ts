import { Directive } from '@angular/core';

@Directive({
  selector: '[rcLine], rc-line',
  host: {
    class: 'rc-line',
  },
})
export class LineDirective {}
