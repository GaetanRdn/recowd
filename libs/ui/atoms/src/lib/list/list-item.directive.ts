import { Directive } from '@angular/core';

@Directive({
  selector: '[rcListItem], rc-list-item',
  host: {
    role: 'list-item',
    class: 'rc-list-item',
  },
})
export class ListItemDirective {}
