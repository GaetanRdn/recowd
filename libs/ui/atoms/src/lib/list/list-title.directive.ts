import { Directive } from '@angular/core';

@Directive({
  selector: '[rcListTitle], rc-list-title',
  host: {
    class: 'rc-list-title',
  },
})
export class ListTitleDirective {}
