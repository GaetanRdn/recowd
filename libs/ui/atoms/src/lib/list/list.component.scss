@use '../assets/utils';

@mixin rc-list($config) {
  $list-item-typo: utils.map-get-deep($config, typography, body-regular);
  $primary: utils.map-get-deep($config, colors, primary);

  rc-list {
    display: flex;
    flex-flow: column;
    gap: 1rem;

    .rc-list-items-wrapper {
      display: flex;
      gap: 0.5rem;
    }

    .rc-list-item,
    .rc-list-title {
      @include utils.use-typography($list-item-typo);
      white-space: nowrap;
    }

    .rc-list-item {
      color: map-get($primary, color);
    }

    .rc-list-title {
      color: darken(map-get($primary, color), 30%);
    }

    .rc-line {
      display: flex;
      align-items: center;

      &:not(:last-child) {
        padding-bottom: 0.5rem;
      }
    }

    &.rc-horizontal {
      .rc-list-item:not(:last-child)::after {
        content: ', ';
      }
    }

    &:not(.rc-horizontal) {
      .rc-list-items-wrapper {
        flex-flow: column;
        gap: 0.8rem;
      }
    }
  }
}
