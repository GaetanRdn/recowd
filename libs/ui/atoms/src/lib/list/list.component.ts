import { ChangeDetectionStrategy, Component, Input } from '@angular/core';

@Component({
  selector: 'rc-list',
  host: {
    role: 'list',
    '[class.rc-horizontal]': 'horizontal',
  },
  template: ` <ng-content select="[rcListItem], rc-list-item"></ng-content>`,
  styleUrls: ['./list.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ListComponent {
  @Input() public horizontal = false;
}
