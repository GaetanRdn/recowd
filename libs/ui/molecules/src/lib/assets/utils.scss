@function map-get-deep($map, $level, $subLevel) {
  @return map-get($map, $level, $subLevel);
}

@mixin use-typography($typography) {
  @each $key in font-size, font-family, font-weight, line-height {
    #{$key}: map-get($typography, $key);
  }
}
