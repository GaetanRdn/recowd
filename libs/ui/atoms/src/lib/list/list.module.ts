import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ListItemDirective } from './list-item.directive';
import { ListComponent } from './list.component';
import { ListTitleDirective } from './list-title.directive';
import { LineDirective } from './line.directive';

@NgModule({
  declarations: [
    ListComponent,
    ListItemDirective,
    ListTitleDirective,
    LineDirective,
  ],
  exports: [
    ListComponent,
    ListItemDirective,
    ListTitleDirective,
    LineDirective,
  ],
  imports: [CommonModule],
})
export class ListModule {}
