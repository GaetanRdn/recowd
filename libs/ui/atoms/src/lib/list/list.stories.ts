import { Meta, moduleMetadata, Story } from '@storybook/angular';
import { ListModule } from './list.module';
import { ListComponent } from './list.component';

export default {
  title: 'atoms/list',
  parameters: {
    jest: 'card.component',
  },
  decorators: [
    moduleMetadata({
      imports: [ListModule],
    }),
  ],
} as Meta<ListComponent>;

const Template: Story<ListComponent> = (args: ListComponent) => ({
  props: {
    horizontal: args.horizontal,
  },
  template: `<rc-list [horizontal]="horizontal">
<rc-list-item>Item 1</rc-list-item>
<rc-list-item>Item 2</rc-list-item>
<rc-list-item>Item 3</rc-list-item>
</rc-list>`,
});

export const Default = Template.bind({});
Default.args = {
  horizontal: false,
};
