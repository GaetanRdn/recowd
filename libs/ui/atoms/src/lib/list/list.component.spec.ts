import { TestBed, waitForAsync } from '@angular/core/testing';
import { Component } from '@angular/core';
import { ListModule } from './list.module';
import { TemplateLookup } from '@recowd/test/utils';

describe('ListComponent', () => {
  let templateLookup: TemplateLookup<HostComponent | HostWithTitleComponent>;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [
        HostComponent,
        HostWithTitleComponent,
        HostWithLineComponent,
      ],
      imports: [ListModule],
    }).compileComponents();
  }));

  it('should create empty', () => {
    templateLookup = new TemplateLookup(HostComponent);
    templateLookup.detectChanges();
    expect(templateLookup.firstChildElement).toMatchSnapshot();
  });

  it('should create with items', () => {
    templateLookup = new TemplateLookup(HostComponent);
    templateLookup.hostComponent.list = ['Item 1', 'Item 2'];

    templateLookup.detectChanges();

    expect(templateLookup.firstChildElement).toMatchSnapshot();
  });

  it('should create horizontal', () => {
    templateLookup = new TemplateLookup(HostComponent);
    (templateLookup.hostComponent as HostComponent).horizontal = true;

    templateLookup.detectChanges();

    expect(templateLookup.firstChildElement).toMatchSnapshot();
  });

  it('should create with title', () => {
    templateLookup = new TemplateLookup(HostWithTitleComponent);

    templateLookup.detectChanges();

    expect(templateLookup.firstChildElement).toMatchSnapshot();
  });

  it('should create with line', () => {
    templateLookup = new TemplateLookup(HostWithLineComponent);

    templateLookup.detectChanges();

    expect(templateLookup.firstChildElement).toMatchSnapshot();
  });
});

@Component({
  template: ` <rc-list [horizontal]="horizontal">
    <rc-list-item *ngFor="let item of list">{{ item }}</rc-list-item>
  </rc-list>`,
})
class HostComponent {
  public list: string[] = [];
  public horizontal = false;
}

@Component({
  template: ` <rc-list>
    <rc-list-title>Title</rc-list-title>
    <rc-list-item *ngFor="let item of list">{{ item }}</rc-list-item>
  </rc-list>`,
})
class HostWithTitleComponent {
  public list: string[] = ['Item'];
}

@Component({
  template: ` <rc-list>
    <rc-list-title>Title</rc-list-title>
    <rc-list-item *ngFor="let item of list">
      <rc-line>{{ item }}</rc-line>
    </rc-list-item>
  </rc-list>`,
})
class HostWithLineComponent {
  public list: string[] = ['Item'];
}
