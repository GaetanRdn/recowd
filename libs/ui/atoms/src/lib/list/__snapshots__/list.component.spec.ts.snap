// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListComponent should create empty 1`] = `
<rc-list
  role="list"
>
  <span
    class="rc-list-items-wrapper"
  >

  </span>
</rc-list>
`;

exports[`ListComponent should create horizontal 1`] = `
<rc-list
  class="rc-horizontal"
  role="list"
>
  <span
    class="rc-list-items-wrapper"
  >

  </span>
</rc-list>
`;

exports[`ListComponent should create with items 1`] = `
<rc-list
  role="list"
>
  <span
    class="rc-list-items-wrapper"
  >
    <rc-list-item
      class="rc-list-item"
      role="list-item"
    >
      Item 1
    </rc-list-item>
    <rc-list-item
      class="rc-list-item"
      role="list-item"
    >
      Item 2
    </rc-list-item>

  </span>
</rc-list>
`;

exports[`ListComponent should create with line 1`] = `
<rc-list
  role="list"
>
  <rc-list-title
    class="rc-list-title"
  >
    Title
  </rc-list-title>
  <span
    class="rc-list-items-wrapper"
  >
    <rc-list-item
      class="rc-list-item"
      role="list-item"
    >
      <rc-line
        class="rc-line"
      >
        Item
      </rc-line>
    </rc-list-item>

  </span>
</rc-list>
`;

exports[`ListComponent should create with title 1`] = `
<rc-list
  role="list"
>
  <rc-list-title
    class="rc-list-title"
  >
    Title
  </rc-list-title>
  <span
    class="rc-list-items-wrapper"
  >
    <rc-list-item
      class="rc-list-item"
      role="list-item"
    >
      Item
    </rc-list-item>

  </span>
</rc-list>
`;

exports[`ListComponent should create with title 1`] = `
<rc-list
  role="list"
>
  <rc-list-title
    class="rc-list-title"
  >
    Title
  </rc-list-title>
  <rc-list-item
    class="rc-list-item"
    role="list-item"
  >
    Item
  </rc-list-item>

</rc-list>
`;
